File: Algo/Leetcode/560SubarraySumEqualToK.cpp
// O(n^2)
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        if(nums.size()==0) return 0;
        int n = nums.size();
        int count = 0;
        for(int i=0; i<n; i++){
            int sum = 0;
            for(int j=i; j>=0; j--){
                sum += nums[j];
                if(sum == k)
                    count ++;
            }
        }
        return count;
    }
};

// O(n). Use a dict to store the sum from before array
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        if(nums.size()==0) return 0;
        int n = nums.size();
        map<int, int> m;
        int sum = 0;
        m[sum] = 1;
        int count = 0;
        for(int i=0; i<n; i++){
            sum += nums[i];
            if(m.find(sum - k) != m.end()) count += m[sum-k];
            m[sum] ++;
        }

        return count;
    }
};

File: Algo/Leetcode/560SubarraySumEqualToK.py
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        m = {0: 1}
        sum = 0
        count = 0
        for n in nums:
            sum += n
            count += m.get(sum-k, 0)
            m[sum] = m.get(sum, 0) + 1

        return count